### ISSUE TITLE
UNSUPPORTED device for HmIP-FSM16

### ISSUE THREAD

**Issue by bensenkuh** [danielperna84/pyhomematic#89]
Any chance to add this component to the supported list?

**Comment by danielperna84** [comment]
Sure. Someone has to do the implementation though, which probably would be you. ;-)
Looking at the product page it seems similar to the HMIP-PSM. Try adding a condition for `HmIP.FSM16` here: https://github.com/danielperna84/pyhomematic/blob/master/pyhomematic/devicetypes/actors.py#L329

**Comment by bensenkuh** [comment]
Yeah, already did and it works. Thank you. :-D

But do i lose it again with the next update?

Currently i played around with hassio but i can't find to change it there... any ideas?

**Comment by danielperna84** [comment]
Yes, you will lose that whenever pyhomematic gets updated. Updating pyhomematic overwrites the actors.py, so your changes will be lost.
You can't modify pyhomematic with hassio. It's sort of _read only_ in that case. So you'll have to wait for an update to get the funcitonality.

Do you want to fork and create a pull request to get your device added, or should I do that quickly?

**Comment by bensenkuh** [comment]
If you have a second for it, that would be great! :-)

Where do you place that kind of requests? Next time i'll could do that on my own ;)

**Comment by danielperna84** [comment]
You simply fork the repository, make the changes, push them to your fork, and then a button will be displayed here which will let you create the PR. :)

**Comment by bensenkuh** [comment]
Sounds easy :-)

First day for me ... thanks for your help!

**Comment by danielperna84** [comment]
Just to confirm: you added it to the condition where it ends up in `sensorIndex = 6`? There is the FSM without the number in the and that is at channel 5. Just want to make sure I use the correct channel for the sensors, and I guess the 16A would just be the rating for the power, so it could be channel 5 as well. The channel in this case is responsible for the sensors (watts, voltage etc.), which show up as their own entities in Home Assistant.

**Comment by bensenkuh** [comment]
Sorry for the late reply but i'm not 100% sure what you mean... ;)

Maybe this will help? 

_2017-10-27 05:16:01 DEBUG (Thread-12) [pyhomematic._hm] RPCFunctions.newDevices: interface_id = homeassistant-ip, dev_descriptions = [{'FLAGS': 1, 'FIRMWARE': '1.6.2', 'AVAILABLE_FIRMWARE': '1.6.2', 'RX_MODE': 0, 'INTERFACE': '', 'SUBTYPE': 'FSM', 'TYPE': 'HmIP-FSM16', 'DIRECTION': 0, 'PARAMSETS': ['MASTER', 'SERVICE'], 'AES_ACTIVE': 1, 'FIRMWARE_UPDATE_STATE': 'UP_TO_DATE', 'ADDRESS': '000B9569A51153', 'ROAMING': 0, 'INDEX': 0, 'VERSION': 2, 'TEAM': '', 'UPDATABLE': True, 'TEAM_TAG': '', 'PARENT_TYPE': '', 'LINK_SOURCE_ROLES': '', 'RF_ADDRESS': 9684117, 'GROUP': '', 'LINK_TARGET_ROLES': '', 'PARENT': '', 'TEAM_CHANNELS': [], 'CHILDREN': ['000B9569A51153:0', '000B9569A51153:1', '000B9569A51153:2', '000B9569A51153:3', '000B9569A51153:4', '000B9569A51153:5', '000B9569A51153:6', '000B9569A51153:7']}, {'FLAGS': 1, 'FIRMWARE': '', 'AVAILABLE_FIRMWARE': '', 'RX_MODE': 0, 'INTERFACE': '', 'SUBTYPE': '', 'TYPE': 'MAINTENANCE', 'DIRECTION': 0, 'PARAMSETS': ['MASTER', 'VALUES', 'SERVICE'], 'AES_ACTIVE': 1, 'FIRMWARE_UPDATE_STATE': '', 'ADDRESS': '000B9569A51153:0', 'ROAMING': 0, 'INDEX': 0, 'VERSION': 2, 'TEAM': '', 'UPDATABLE': True, 'TEAM_TAG': '', 'PARENT_TYPE': 'HmIP-FSM16', 'LINK_SOURCE_ROLES': '', 'RF_ADDRESS': 0, 'GROUP': '', 'LINK_TARGET_ROLES': '', 'PARENT': '000B9569A51153', 'TEAM_CHANNELS': [], 'CHILDREN': []}, {'FLAGS': 1, 'FIRMWARE': '', 'AVAILABLE_FIRMWARE': '', 'RX_MODE': 0, 'INTERFACE': '', 'SUBTYPE': '', 'TYPE': 'SWITCH_TRANSMITTER', 'DIRECTION': 0, 'PARAMSETS': ['MASTER', 'VALUES', 'SERVICE'], 'AES_ACTIVE': 1, 'FIRMWARE_UPDATE_STATE': '', 'ADDRESS': '000B9569A51153:1', 'ROAMING': 0, 'INDEX': 1, 'VERSION': 2, 'TEAM': '', 'UPDATABLE': True, 'TEAM_TAG': '', 'PARENT_TYPE': 'HmIP-FSM16', 'LINK_SOURCE_ROLES': '', 'RF_ADDRESS': 0, 'GROUP': '', 'LINK_TARGET_ROLES': '', 'PARENT': '000B9569A51153', 'TEAM_CHANNELS': [], 'CHILDREN': []}, {'FLAGS': 1, 'FIRMWARE': '', 'AVAILABLE_FIRMWARE': '', 'RX_MODE': 0, 'INTERFACE': '', 'SUBTYPE': '', 'TYPE': 'SWITCH_VIRTUAL_RECEIVER', 'DIRECTION': 2, 'PARAMSETS': ['MASTER', 'VALUES', 'LINK', 'SERVICE'], 'AES_ACTIVE': 1, 'FIRMWARE_UPDATE_STATE': '', 'ADDRESS': '000B9569A51153:2', 'ROAMING': 0, 'INDEX': 2, 'VERSION': 2, 'TEAM': '', 'UPDATABLE': True, 'TEAM_TAG': '', 'PARENT_TYPE': 'HmIP-FSM16', 'LINK_SOURCE_ROLES': '', 'RF_ADDRESS': 0, 'GROUP': '', 'LINK_TARGET_ROLES': 'CONDITIONAL_SWITCH LEVEL SWITCH REMOTE_CONTROL', 'PARENT': '000B9569A51153', 'TEAM_CHANNELS': [], 'CHILDREN': []}, {'FLAGS': 1, 'FIRMWARE': '', 'AVAILABLE_FIRMWARE': '', 'RX_MODE': 0, 'INTERFACE': '', 'SUBTYPE': '', 'TYPE': 'SWITCH_VIRTUAL_RECEIVER', 'DIRECTION': 2, 'PARAMSETS': ['MASTER', 'VALUES', 'LINK', 'SERVICE'], 'AES_ACTIVE': 1, 'FIRMWARE_UPDATE_STATE': '', 'ADDRESS': '000B9569A51153:3', 'ROAMING': 0, 'INDEX': 3, 'VERSION': 2, 'TEAM': '', 'UPDATABLE': True, 'TEAM_TAG': '', 'PARENT_TYPE': 'HmIP-FSM16', 'LINK_SOURCE_ROLES': '', 'RF_ADDRESS': 0, 'GROUP': '', 'LINK_TARGET_ROLES': 'CONDITIONAL_SWITCH LEVEL SWITCH REMOTE_CONTROL', 'PARENT': '000B9569A51153', 'TEAM_CHANNELS': [], 'CHILDREN': []}, {'FLAGS': 1, 'FIRMWARE': '', 'AVAILABLE_FIRMWARE': '', 'RX_MODE': 0, 'INTERFACE': '', 'SUBTYPE': '', 'TYPE': 'SWITCH_VIRTUAL_RECEIVER', 'DIRECTION': 2, 'PARAMSETS': ['MASTER', 'VALUES', 'LINK', 'SERVICE'], 'AES_ACTIVE': 1, 'FIRMWARE_UPDATE_STATE': '', 'ADDRESS': '000B9569A51153:4', 'ROAMING': 0, 'INDEX': 4, 'VERSION': 2, 'TEAM': '', 'UPDATABLE': True, 'TEAM_TAG': '', 'PARENT_TYPE': 'HmIP-FSM16', 'LINK_SOURCE_ROLES': '', 'RF_ADDRESS': 0, 'GROUP': '', 'LINK_TARGET_ROLES': 'CONDITIONAL_SWITCH LEVEL SWITCH REMOTE_CONTROL', 'PARENT': '000B9569A51153', 'TEAM_CHANNELS': [], 'CHILDREN': []}, {'FLAGS': 1, 'FIRMWARE': '', 'AVAILABLE_FIRMWARE': '', 'RX_MODE': 0, 'INTERFACE': '', 'SUBTYPE': '', 'TYPE': 'ENERGIE_METER_TRANSMITTER', 'DIRECTION': 0, 'PARAMSETS': ['MASTER', 'VALUES', 'SERVICE'], 'AES_ACTIVE': 1, 'FIRMWARE_UPDATE_STATE': '', 'ADDRESS': '000B9569A51153:5', 'ROAMING': 0, 'INDEX': 5, 'VERSION': 2, 'TEAM': '', 'UPDATABLE': True, 'TEAM_TAG': '', 'PARENT_TYPE': 'HmIP-FSM16', 'LINK_SOURCE_ROLES': '', 'RF_ADDRESS': 0, 'GROUP': '', 'LINK_TARGET_ROLES': '', 'PARENT': '000B9569A51153', 'TEAM_CHANNELS': [], 'CHILDREN': []}, {'FLAGS': 1, 'FIRMWARE': '', 'AVAILABLE_FIRMWARE': '', 'RX_MODE': 0, 'INTERFACE': '', 'SUBTYPE': '', 'TYPE': 'COND_SWITCH_TRANSMITTER', 'DIRECTION': 1, 'PARAMSETS': ['MASTER', 'VALUES', 'LINK', 'SERVICE'], 'AES_ACTIVE': 1, 'FIRMWARE_UPDATE_STATE': '', 'ADDRESS': '000B9569A51153:6', 'ROAMING': 0, 'INDEX': 6, 'VERSION': 2, 'TEAM': '', 'UPDATABLE': True, 'TEAM_TAG': '', 'PARENT_TYPE': 'HmIP-FSM16', 'LINK_SOURCE_ROLES': 'CONDITIONAL_SWITCH', 'RF_ADDRESS': 0, 'GROUP': '', 'LINK_TARGET_ROLES': '', 'PARENT': '000B9569A51153', 'TEAM_CHANNELS': [], 'CHILDREN': []}, {'FLAGS': 1, 'FIRMWARE': '', 'AVAILABLE_FIRMWARE': '', 'RX_MODE': 0, 'INTERFACE': '', 'SUBTYPE': '', 'TYPE': 'SWITCH_WEEK_PROFILE', 'DIRECTION': 0, 'PARAMSETS': ['MASTER', 'VALUES', 'SERVICE'], 'AES_ACTIVE': 1, 'FIRMWARE_UPDATE_STATE': '', 'ADDRESS': '000B9569A51153:7', 'ROAMING': 0, 'INDEX': 7, 'VERSION': 2, 'TEAM': '', 'UPDATABLE': True, 'TEAM_TAG': '', 'PARENT_TYPE': 'HmIP-FSM16', 'LINK_SOURCE_ROLES': '', 'RF_ADDRESS': 0, 'GROUP': '', 'LINK_TARGET_ROLES': '', 'PARENT': '000B9569A51153', 'TEAM_CHANNELS': [], 'CHILDREN': []}, {'FLAGS': 1, 'FIRMWARE': '1.0.11', 'AVAILABLE_FIRMWARE': '0.0.0', 'RX_MODE': 3, 'INTERFACE': '', 'SUBTYPE': 'SD', 'TYPE': 'HmIP-SWSD', 'DIRECTION': 0, 'PARAMSETS': ['MASTER', 'SERVICE'], 'AES_ACTIVE': 1, 'FIRMWARE_UPDATE_STATE': 'UP_TO_DATE', 'ADDRESS': '000A570994E399', 'ROAMING': 0, 'INDEX': 0, 'VERSION': 1, 'TEAM': '', 'UPDATABLE': True, 'TEAM_TAG': '', 'PARENT_TYPE': '', 'LINK_SOURCE_ROLES': '', 'RF_ADDRESS': 2349364, 'GROUP': '', 'LINK_TARGET_ROLES': '', 'PARENT': '', 'TEAM_CHANNELS': [], 'CHILDREN': ['000A570994E399:0', '000A570994E399:1']}, {'FLAGS': 1, 'FIRMWARE': '', 'AVAILABLE_FIRMWARE': '', 'RX_MODE': 0, 'INTERFACE': '', 'SUBTYPE': '', 'TYPE': 'MAINTENANCE', 'DIRECTION': 0, 'PARAMSETS': ['MASTER', 'VALUES', 'SERVICE'], 'AES_ACTIVE': 1, 'FIRMWARE_UPDATE_STATE': '', 'ADDRESS': '000A570994E399:0', 'ROAMING': 0, 'INDEX': 0, 'VERSION': 1, 'TEAM': '', 'UPDATABLE': True, 'TEAM_TAG': '', 'PARENT_TYPE': 'HmIP-SWSD', 'LINK_SOURCE_ROLES': '', 'RF_ADDRESS': 0, 'GROUP': '', 'LINK_TARGET_ROLES': '', 'PARENT': '000A570994E399', 'TEAM_CHANNELS': [], 'CHILDREN': []}, {'FLAGS': 1, 'FIRMWARE': '', 'AVAILABLE_FIRMWARE': '', 'RX_MODE': 0, 'INTERFACE': '', 'SUBTYPE': '', 'TYPE': 'SMOKE_DETECTOR', 'DIRECTION': 0, 'PARAMSETS': ['MASTER', 'VALUES', 'SERVICE'], 'AES_ACTIVE': 1, 'FIRMWARE_UPDATE_STATE': '', 'ADDRESS': '000A570994E399:1', 'ROAMING': 0, 'INDEX': 1, 'VERSION': 1, 'TEAM': '', 'UPDATABLE': True, 'TEAM_TAG': '', 'PARENT_TYPE': 'HmIP-SWSD', 'LINK_SOURCE_ROLES': '', 'RF_ADDRESS': 0, 'GROUP': '', 'LINK_TARGET_ROLES': '', 'PARENT': '000A570994E399', 'TEAM_CHANNELS': [], 'CHILDREN': []}]_

**Comment by danielperna84** [comment]
No, that does not help because it doesn't show which information is available at which channel.

I am talking about https://github.com/danielperna84/pyhomematic/blob/72a2b3013201ffa779c1d4c190a85d24846b6c39/pyhomematic/devicetypes/actors.py#L327
We already support the FSM-model without the `16` appended. For it we have `sensorIndex = 5`.
In my first answer I proposed adding it here: https://github.com/danielperna84/pyhomematic/blob/72a2b3013201ffa779c1d4c190a85d24846b6c39/pyhomematic/devicetypes/actors.py#L329
That however would result in `sensorIndex = 6`. It has to be one of them, and I am not sure that when testing the functionality you checked if the sensors (power usage) are being displayed. It will work as a switch either way, but in one case you get extra entitys for power etc., in the other you don't.
By looking at the name, the `16` seems to be just a version of the same actor that can handle more power. Hence It should actually NOT be the line I have propesed earlier, but rather the one where the existing `FSM` is, since I wouldn't expect that eq3 would use different channels for an otherwise almost exactly identical device.

So my question is: how does the code look like that lets you control the switch AND get the sensors for power?
